src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/Whitelist.java

```diff
@@ -69,12 +71,18 @@
      * @return an aggregated default list
      */
     public static synchronized @Nonnull Whitelist all() {
+        Jenkins j = Jenkins.getInstance();
+        if (j == null) {
+            return new ProxyWhitelist();
+        }
+        Whitelist all = allByJenkins.get(j);
         if (all == null) {
             // TODO should check for dynamic changes in this list, e.g. from dynamically loaded plugins, and return a fresh aggregate
-            all = new ProxyWhitelist(Jenkins.getInstance().getExtensionList(Whitelist.class));
+            all = new ProxyWhitelist(j.getExtensionList(Whitelist.class));
+            allByJenkins.put(j, all);
         }
         return all;
     }
-    private static Whitelist all;
+    private static final Map<Jenkins,Whitelist> allByJenkins = new WeakHashMap<Jenkins,Whitelist>();
 
```

Comment:
That is what I tried first, but it did not work. I had an `@Extension` which was `AllHolder` with a field corresponding to the `ProxyWhitelist`, but neither `Jenkins.getInstance().getExtensionList(AllHolder.class).get(AllHolder.class)` nor `Jenkins.lookup(AllHolder.class)` returned anything; perhaps because `AllHolder` was not extending any interface so it was not being categorized as anything.

I also considered registering a `ProxyWhitelist` subclass, which solves that problem, but then introduces a different problem that the extension list now includes itself, which I would need to filter out.

So just using a `WeakHashMap` seemed simpler than fighting these issues.